dcap/types/src/quote_header.rs

```diff
@@ -54,9 +55,88 @@ new_type_accessors_impls! {
     QuoteHeader, sgx_quote_header_t;
 }
 
+impl From<[u8; mem::size_of::<sgx_quote_header_t>()]> for QuoteHeader {
+    fn from(bytes: [u8; mem::size_of::<sgx_quote_header_t>()]) -> Self {
+        // A note about the `expect()` calls.  This size is specified in the
+        // signature and there are unit tests ensuring the extraction from a
+        // byte array.  The values should not fail to extract due to size
+        // issues.
+        let mut header = Self::default();
+        header.0.version =
+            u16::from_le_bytes(bytes[..2].try_into().expect("Failed to extract `version`"));
+        header.0.att_key_type = u16::from_le_bytes(
+            bytes[2..4]
```

Comment:
I've been thinking about magic numbers like these and in use in ReportBody's `From` implementation.

I'm not fond of making a bunch of defines for all these value when there only used here as it creates  a level of indirection and doesn't ensure correctness.
One thought I had to help address this is to come up with some kind of incrementable range:
```rust
let range = Magic::new(0, mem::sizeof::<u16>());
header.0.att_key_type = u16::from_le_bytes(bytes[range].try_into()..expect("blah");
let range = Magic::new_from_other(range, mem::sizeof::<u32>())
header.0.att_key_data_0 = u32::from_le_bytes(bytes[range].try_into()..expect("blah");
```
IDK it feels a bit clunky, but thought I would throw it out there in case someone has thoughts on:
- ways to make it more palatable
- arguments exposing the wrongness of this idea

Reply:
I think for this specific object, there's no reason to worry about validating the data, since that decision can be punted until the user actually tries to create a versioned quote object from the bytes underlying this structure.

What I would suggest is flipping this problem on it's head: each `QuoteN` variant actually implements a trait that contains what this magic number should be, e.g.:

```rust
pub trait VersionedQuote {
    const VERSION: u16;
    type Err: Display;

    fn try_from_header(header: &QuoteHeader, bytes: &[u8]) -> Result<Self, Self::Error> {
    }
}
```

Reply:
Oh, I see, you're talking about the byte ranges...

Is this something `nom` could solve?

Reply:
nom may be the way to go.
That's probably a better way to think of it; a consumed stream or a cursor of some kind

Reply:
I've created a task on #66 to pull in nom.  I think it can be done as a follow-on PR.